colossus-metrics/src/main/scala/colossus/metrics/Collection.scala

```diff
@@ -117,29 +121,72 @@ class Collection(val config: CollectorConfig) {
         )
       }
     }
-    if (collectors.containsKey(address)) {
-      cast(collectors.get(address))
-    } else {
+    addContext(ns)
+    get(address).fold{
       val c = f(address, config)
-      collectors.putIfAbsent(address, c) match {
-        case null => c
-        case other => cast(other)
+      put(address, c) match {
+        case None => c
+        case Some(other) => cast(other)
       }
+    }{retrieved =>
+      cast(retrieved)
     }
   }
 
   def tick(interval: FiniteDuration): MetricMap = {
-    val keys = collectors.keys
     var build: MetricMap = Map()
-    while (keys.hasMoreElements) {
-      val key = keys.nextElement
-      Option(collectors.get(key)) foreach { c =>
-        build = build ++ c.tick(interval)
+    iteratorWithTags.foreach{ case (address, extraTags, c) =>
+      build = build ++ c.tick(interval).mapValues{ valueMap =>
+        valueMap.map{case (t, value) => (t ++ extraTags, value)}
       }
     }
     build
   }
 
+  def withPrefix(s: MetricAddress): Collection =
+    if (s.size == 0) {
+      this
+    } else {
+      val leading = MetricAddress(s.head)
+      suffixes.putIfAbsent(leading, empty)
+      suffixes.get(leading).withPrefix(s.tail)
+    }
+  
+  def addContext(ns: MetricNamespace) = withPrefix(ns.namespace).tags = ns.extraTags
+  
+  override def update(s: MetricAddress, elem: Collector) =
+    withPrefix(s).collector = Some(elem)
+
+  override def +=(kv: (MetricAddress, Collector)): Collection.this.type = { update(kv._1, kv._2); this }
+  
+  override def -=(key: MetricAddress): Collection.this.type = { throw new UnsupportedOperationException }
+
+  override def get(key: MetricAddress): Option[Collector] = {
+    if (key.size == 0) {
+      collector
+    } else {
+      if(suffixes.containsKey(MetricAddress(key.head))) {
+        suffixes.get(MetricAddress(key.head)).get(key.tail)
+      } else {
+        None
+      }
+    }
+  }
+
+  override def iterator: Iterator[(MetricAddress, Collector)] = (for (v <- collector.iterator) yield (MetricAddress.Root, v)) ++
```

Comment:
`MetricAddress.Root` is correct here, what is returned here is appended to end of previous call creating the whole metric address.